src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityPowerSubstation.java

```diff
@@ -0,0 +1,638 @@
+        inputs.addAll(getAbilities(MultiblockAbility.INPUT_ENERGY));
+        inputs.addAll(getAbilities(MultiblockAbility.SUBSTATION_INPUT_ENERGY));
+        this.inputHatches = new EnergyContainerList(inputs);
+
+        List<IEnergyContainer> outputs = new ArrayList<>();
+        outputs.addAll(getAbilities(MultiblockAbility.OUTPUT_ENERGY));
+        outputs.addAll(getAbilities(MultiblockAbility.SUBSTATION_OUTPUT_ENERGY));
+        this.outputHatches = new EnergyContainerList(outputs);
+
+        List<IBatteryData> parts = new ArrayList<>();
+        for (Map.Entry<String, Object> battery : context.entrySet()) {
+            if (battery.getKey().startsWith(PMC_BATTERY_HEADER)) {
+                BatteryMatchWrapper wrapper = (BatteryMatchWrapper) battery.getValue();
+                for (int i = 0; i < wrapper.amount; i++) {
+                    parts.add(wrapper.partType);
+                }
+            }
+        }
+        if (parts.isEmpty()) {
+            // only empty batteries found in the structure
+            invalidateStructure();
+            return;
+        }
+        if (this.energyBank == null) {
+            this.energyBank = new PowerStationEnergyBank(parts);
+        } else {
+            this.energyBank = energyBank.rebuild(parts);
+        }
+        this.passiveDrain = this.energyBank.getPassiveDrainPerTick();
+    }
+
+    @Override
+    public void invalidateStructure() {
+        // don't null out energyBank since it holds the stored energy, which
+        // we need to hold on to across rebuilds to not void all energy if a
+        // multiblock part or block other than the controller is broken.
+        inputHatches = null;
+        outputHatches = null;
+        passiveDrain = 0;
+        totalIOLastSec = 0;
+        averageIOLastSec = 0;
+        super.invalidateStructure();
+    }
+
+    @Override
+    protected void updateFormedValid() {
+        if (!getWorld().isRemote) {
+            if (getOffsetTimer() % 20 == 0) {
+                // active here is just used for rendering
+                setActive(energyBank.hasEnergy());
+                averageIOLastSec = totalIOLastSec / 20;
+                totalIOLastSec = 0;
+            }
+
+            if (isWorkingEnabled()) {
+                // Bank from Energy Input Hatches
+                long energyBanked = energyBank.fill(inputHatches.getEnergyStored());
+                inputHatches.changeEnergy(-energyBanked);
+                totalIOLastSec += energyBanked;
+
+                // Passive drain
+                long energyPassiveDrained = energyBank.drain(getPassiveDrain());
+                totalIOLastSec -= energyPassiveDrained;
+
+                // Debank to Dynamo Hatches
+                long energyDebanked = energyBank.drain(outputHatches.getEnergyCapacity() - outputHatches.getEnergyStored());
+                outputHatches.changeEnergy(energyDebanked);
+                totalIOLastSec -= energyDebanked;
```

Comment:
Is subtracting here correct? Since you want to find the total IO, shouldn't that be the sum of what is inserted and what is extracted?

Reply:
I guess more accurately that variable is the net IO over the last second, so I'll rename it